src/server/handlers/consents/{ID}/generateChallenge.ts

```
@@ -0,0 +1,130 @@
+} from '../../../../domain/consents/{ID}/generateChallenge'
+import * as challenge from '../../../../lib/challenge'
+import { Request, ResponseToolkit, ResponseObject } from '@hapi/hapi'
+import { consentDB, scopeDB } from '../../../../lib/db'
+import { Consent } from '../../../../model/consent'
+import Logger from '@mojaloop/central-services-logger'
+import { convertScopesToExternal } from '../../../../lib/scopes'
+import { Scope } from '../../../../model/scope'
+
+// Asynchronously deals with generating challenge, updating consent db
+//  and making outgoing PUT consent/{ID} call
+export async function generateChallengeAndPutConsentId (
+  request: Request,
+  consent: Consent,
+  id: string
+): Promise<void> {
+  try {
+    // If there is no pre-existing challenge for the consent id
+    // Generate one and update database
+    if (!consent.credentialChallenge) {
+      // Challenge generation
+      const challengeValue = await challenge.generate()
+
+      // Updating credentials with generated challenge
+      const credential: ConsentCredential = {
+        credentialChallenge: challengeValue,
+        credentialStatus: 'PENDING',
+        credentialType: 'FIDO'
+      }
+
+      consent = await updateConsentCredential(consent, credential)
+    } else if (consent.credentialStatus === 'ACTIVE') {
+      // TODO: Error handling here - dealt with in #355
+      Logger.error('ACTIVE credential consent has requested challenge')
+      return
+    }
+
+    // Retrieve Scopes
+    const scopesRetrieved: Scope[] = await scopeDB.retrieveAll(id)
+    const scopes = convertScopesToExternal(scopesRetrieved)
+
+    // Outgoing call to PUT consents/{ID}
+    await putConsentId(consent, request, scopes)
+  } catch (error) {
+    Logger.push(error)
+    // eslint-disable-next-line max-len
+    Logger.error(`Error: Outgoing call with challenge credential NOT made to  PUT consent/${id}`)
+    // TODO: Decide on error handling HERE -  dealt with in future ticket #355
```

Let rethrow the received error instead of keeping it silent. In the next story we will transfer it to PUT .../error

---

@eoln if we throw an error here or in the previous comment - there will be an unhandled promise rejection issue as nothing is handling the POST handler's errors

---

Yes, you are right, normally we shouldn't allow promise rejections to bubble up, but unhandled promise rejection will be temporary - until we got PUT .../error implemented. IMHO it is better to have comment TODO in code, and also unhandled promise rejection/exception which will help us remember that we have some unfinished flow

---

makes sense have addressed this